Add basic support for memory pool

doru1004 opened the pull request:
This patch enables the use of a memory pool for allocating the internal MemRefs used by an ONNX model.

The patch introduces a new op:

```
krnl.getref
```

which takes two arguments: the memory pool and the offset inside the memory where the data of the MemRef that is returned begins.

This new operation is meant to replace the current way of allocating internal MemRefs:

```
%0 = alloc() : memref<10x10xf32>
```

In the case of the alloca above, the krnl.getref transformed code would look like this:

```
%c0_i64 = constant 0 : i64
%0 = alloc() : memref<400xi8>
%1 = "krnl.getref"(%0, %c0_i64) : (memref<400xi8>, i64) -> memref<10x10xf32>
```

In LLVM this will translate to:

```
    ;
    ;  constant 0 : i64
    ;
    %15 = llvm.mlir.constant(0 : i64) : !llvm.i64
    ...

    ;
    ; alloc() : memref<400xi8>
    ;
    %37 = llvm.mlir.constant(400 : index) : !llvm.i64
    %38 = llvm.mlir.null : !llvm<"i8*">
    %39 = llvm.mlir.constant(1 : index) : !llvm.i64
    %40 = llvm.getelementptr %38[%39] : (!llvm<"i8*">, !llvm.i64) -> !llvm<"i8*">
    %41 = llvm.ptrtoint %40 : !llvm<"i8*"> to !llvm.i64
    %42 = llvm.mul %37, %41 : !llvm.i64
    %43 = llvm.call @malloc(%42) : (!llvm.i64) -> !llvm<"i8*">
    %44 = llvm.bitcast %43 : !llvm<"i8*"> to !llvm<"i8*">
    %45 = llvm.mlir.undef : !llvm<"{ i8*, i8*, i64, [1 x i64], [1 x i64] }">
    %46 = llvm.insertvalue %44, %45[0] : !llvm<"{ i8*, i8*, i64, [1 x i64], [1 x i64] }">
    %47 = llvm.insertvalue %44, %46[1] : !llvm<"{ i8*, i8*, i64, [1 x i64], [1 x i64] }">
    %48 = llvm.mlir.constant(0 : index) : !llvm.i64
    %49 = llvm.insertvalue %48, %47[2] : !llvm<"{ i8*, i8*, i64, [1 x i64], [1 x i64] }">
    %50 = llvm.mlir.constant(1 : index) : !llvm.i64
    %51 = llvm.insertvalue %37, %49[3, 0] : !llvm<"{ i8*, i8*, i64, [1 x i64], [1 x i64] }">
    %52 = llvm.insertvalue %50, %51[4, 0] : !llvm<"{ i8*, i8*, i64, [1 x i64], [1 x i64] }">

    ;
    ; "krnl.getref"(%0, %c0_i64) : (memref<400xi8>, i64) -> memref<10x10xf32>
    ;
    %53 = llvm.extractvalue %52[1] : !llvm<"{ i8*, i8*, i64, [1 x i64], [1 x i64] }">
    %54 = llvm.getelementptr %53[%15] : (!llvm<"i8*">, !llvm.i64) -> !llvm<"i8*">
    %55 = llvm.bitcast %54 : !llvm<"i8*"> to !llvm<"float*">
    %56 = llvm.mlir.undef : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
    %57 = llvm.insertvalue %55, %56[0] : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
    %58 = llvm.insertvalue %55, %57[1] : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
    %59 = llvm.mlir.constant(0 : index) : !llvm.i64
    %60 = llvm.insertvalue %59, %58[2] : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
    %61 = llvm.mlir.constant(10 : index) : !llvm.i64
    %62 = llvm.insertvalue %61, %60[3, 0] : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
    %63 = llvm.mlir.constant(10 : index) : !llvm.i64
    %64 = llvm.insertvalue %63, %62[4, 0] : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
    %65 = llvm.mlir.constant(10 : index) : !llvm.i64
    %66 = llvm.insertvalue %65, %64[3, 1] : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
    %67 = llvm.mlir.constant(1 : index) : !llvm.i64
    %68 = llvm.insertvalue %67, %66[4, 1] : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
```

Currently each internal allocated MemRef has its own memory pool. Next step is to bundle all the internal memory pools into a single memory pool and generate offsets into this memory pool for the different `getref` instructions.

Memory pooling only applies to statically shaped MemRefs. In the future we plan to extend this functionality to dynamic MemRefs.

doru1004 commented on src/Transform/EnableMemoryPool.cpp:
```diff
@@ -0,0 +1,159 @@
+      return failure();
+
+    // Check the result of this alloc is not already used by a krnl.getref.
+    if (checkOpResultIsUsedByGetRef(&allocOp))
```
This only looks for the allocs which can be memory pooled. Not all Allocs can be placed in the memory pool. The one exception is the alloc which we return.

doru1004 commented on src/Dialect/Krnl/KrnlOps.td:
```diff
@@ -205,3 +205,16 @@ def KrnlGlobalOp : Op<Krnl_Dialect, "global"> {
+def KrnlGetRefOp : Op<Krnl_Dialect, "getref"> {
+  let summary = "Krnl a MemRef from within another MemRef starting at a specific offset.";
+  let description = [{
+    Retreieves a MemRef from within another MemRef.
```
> I am a bit concerned when everything we do is with constant sizes... it will make the support of dyn memory harder. I let you judge if we should charge ahead with this PR first and then add it, or if we should add it from the start.
> 
> Otherwise, please add comments to the op definitions.

For now we only support static sizes but dynamic support can be added. I recommend adding dynamic support in a separate patch since this patch is already big enough.

doru1004 replied:
There's no restriction as to what types you can use for the input memref. The offset is always going to be in the unit of the first argument. it is up to the "user" of the krnl.getref instruction to set the offset to be a valid offset within the first MemRef argument.

In our case what happens is that we have a transformation which takes in an alloc operation and transforms it into a correct alloc + getref instruction combination. We emit a flat MemRef of i8 type and an offset to match.

Yes the offset can be more than just a constant but we don't yet have a use case for that. In order to emit the offset as not a constant we would have to teach onnx-mlir to emit an expression for this offset. When that happens it will be fine to pass it to the getref instruction.

doru1004 commented:
> I am a bit concerned when everything we do is with constant sizes... it will make the support of dyn memory harder. I let you judge if we should charge ahead with this PR first and then add it, or if we should add it from the start.
> 
> Otherwise, please add comments to the op definitions.

For now we only support static sizes but dynamic support can be added. I recommend adding dynamic support in a separate patch since this patch is already big enough.

AlexandreEichenberger commented:
it's fine, maybe we will do a list of places to add it, let's use a "todo dyn shape" comment to start labeling places to add it.

Please add comments for the new ops, and I will then approve the patch, thanks

doru1004 commented:
> it's fine, maybe we will do a list of places to add it, let's use a "todo dyn shape" comment to start labeling places to add it.
> 
> Please add comments for the new ops, and I will then approve the patch, thanks

The TODO exists already.

Not sure what comment you'd like for the new operation I added.

doru1004 commented:
@AlexandreEichenberger I added the additional comment.

I also added a new test which showcases that `krnl.getref` is generic and that it works with MemRefs of any type.
